ecommerce/views_test.py

```diff
@@ -1227,6 +1227,27 @@ def test_products_viewset_list(user_drf_client, coupon_product_ids):
         )
 
 
+def test_products_viewset_list_ordering(user_drf_client):
+    """
+    Test that the ProductViewSet returns all products ordered alphabetically
+    by course/program title
+    """
+    response = user_drf_client.get(reverse("products_api-list"))
+    assert response.status_code == status.HTTP_200_OK
+    products = response.json()
+    serialized_products_list = ProductSerializer(data=products, many=True).initial_data
+    ordered_products_title = sorted(
+        serialized_products_list,
+        key=lambda k: k["content_object"]["course"]["title"]
+        if "course" in k["content_object"]
+        else k["content_object"]["title"],
+    )
+    ordered_products_type = sorted(
+        ordered_products_title, key=lambda k: k["product_type"], reverse=True
+    )
+    assert ordered_products_type == serialized_products_list
```

Comment:
I don't think this was testing anything at all. I set a breakpoint in this test case b/c I didn't see where the test data could be coming from. This is what I saw:

```
[30] > /src/ecommerce/views_test.py(1239)test_products_viewset_list_ordering()
-> serialized_products_list = ProductSerializer(data=products, many=True).initial_data
   5 frames hidden (try 'help hidden_frames')
(Pdb++) products
[]
(Pdb++) n
[30] > /src/ecommerce/views_test.py(1240)test_products_viewset_list_ordering()
-> ordered_products_title = sorted(
   5 frames hidden (try 'help hidden_frames')
(Pdb++) serialized_products_list
[]
```

Generally speaking, it's best if you do one of two things: (1) set up the test data directly in the test function using factories, or (2) set up test data in a pytest fixture, then include that fixture explicitly in the test function arguments ([pytest fixture reference](https://docs.pytest.org/en/stable/fixture.html), example from our codebase: [fixture definition](https://github.com/mitodl/mitxpro/blob/master/cms/views_test.py#L34), [usage](https://github.com/mitodl/mitxpro/blob/master/cms/views_test.py#L42)). I know there are some patterns that don't exactly follow that in our codebases, but please follow that practice whenever you can.

Suggested fix below. This will work if you add `import factory` to the imports.

```suggestion
    runs = CourseRunFactory.create_batch(2, title=factory.Iterator(["Z Product", "A Product"]))
    programs = ProgramFactory.create_batch(2, title=factory.Iterator(["Z Product", "A Product"]))
    ProductVersionFactory.create_batch(4, product__content_object=factory.Iterator(runs + programs))
    response = user_drf_client.get(reverse("products_api-list"))
    assert response.status_code == status.HTTP_200_OK
    products = response.json()
    product_object_ids = [product["latest_version"]["object_id"] for product in products]
    # Results should be returned with programs before course runs, then alphabetized by title
    assert product_object_ids == [
        programs[1].id,
        programs[0].id,
        runs[1].id,
        runs[0].id,
    ]
    assert [product["latest_version"]["type"] for product in products] == ["program", "program", "courserun", "courserun"]
```

Reply:
Thanks for your suggestion. 

I incorporated the feedback with the suggested fix. Though I noticed that this fix in the test was showing inconsistent behavior and digging further I noticed that our course titles were generating randomly through CourseFactory([Ref](https://github.com/mitodl/mitxpro/blob/master/courses/factories.py)) when we created batch using CourseRunFactory([Ref](https://github.com/mitodl/mitxpro/blob/master/courses/factories.py)), because we were looking for course titles to be alphabetized instead of `CourseRuns's` titles so the test was showing inconsistent behavior. 

I made minor changes on how was test data created and changed the titles for a little more readability. Please let me know what you think about it.